cmd/proxy/actions/app_proxy.go

```diff
@@ -18,11 +20,36 @@ func addProxyRoutes(
 	app.GET("/healthz", healthHandler)
 
 	// Download Protocol
+	// the download.Protocol and the stash.Stasher interfaces are composable
+	// in a middleware fashion. Therefore you can separate concerns
+	// by the functionality: a download.Protocol that just takes care
+	// of "go getting" things, and another Protocol that just takes care
+	// of "pooling" requests etc.
+
+	// In our case, we'd like to compose both interfaces in a particular
+	// order to ensure logical ordering of execution.
+
+	// Here's the order of an incoming request to the download.Protocol:
+
+	// 1. The downloadpool gets hit first, and manages concurrent requests
+	// 2. The downloadpool passes the request to its parent Protocol: stasher
+	// 3. The stasher Protocol checks storage first, and if storage is empty
+	// it makes a Stash request to the stash.Stasher interface.
+
+	// Once the stasher picks up an order, here's how the requests go in order:
+	// 1. The singleflight picks up the first request and latches duplicate ones.
+	// 2. The singleflight passes the stash to its parent: stashpool.
+	// 3. The stashpool manages limiting concurrent requests and passes them to stash.
+	// 4. The plain stash.New just takes a request from upstream and saves it into storage.
 	gg, err := goget.New()
 	if err != nil {
 		return err
 	}
-	p := download.New(gg, s, env.GoGetWorkers())
+	st := stash.New(gg, s, stash.WithSingleflight, stash.WithPool(env.GoGetWorkers()))
```

Comment:
@marwan-at-work do you think it would make sense to split the DP interface into smaller ones or just pass functions into `stash.New(...` and `addons.WithStasher(...)`? I'm asking because it's a little confusing to me when I see that `stash.New` takes DP interface because I know it only needs one function - `Version`
The same with `addons.WithStasher` - it wasn't clear to me why it needs the DP at all. Then I looked up and it's obviously needed for `List` and `Latest` but again, I had to look up deeper to see what's going on.
Let me know what you think.

Reply:
@marpio updated the PR. Instead of creating yet another type, we can just use the `module.Fetcher` interface. However, that interface can use some refactoring to make this PR cleaner. But I don't want to go that far. In another PR, I would make the module.Fetcher.Fetch method just immediately return a `*storage.Version` instead of having it be two steps (Fetch and Ref)

Reply:
:+1:  I was also thinking about pulling `module.Fetcher' out of goget